File: LeetCode/105.Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal/Mario/105.Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.py
#!/usr/bin/env python3


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        dic = {val: i for i, val in enumerate(inorder)}

        def helper(pre_l, pre_r, in_l, in_r):
            # base case
            if pre_l > pre_r:
                return None

            val = preorder[pre_l]
            root = TreeNode(val)

            idx = dic[val]
            left_size = idx - in_l

            root.left = helper(pre_l + 1, pre_l + left_size, in_l, idx - 1)
            root.right = helper(pre_l + 1 + left_size, pre_r, idx + 1, in_r)

            return root

        n = len(preorder)
        return helper(0, n - 1, 0, n - 1)

File: LeetCode/105.Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal/105.Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {

    private Map<Integer, Integer> map;
    private int index;

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        index = 0;
        map = new HashMap<>();
        for (int i = 0; i < inorder.length; i++) {
            map.put(inorder[i], i);
        }
        return dfs(preorder, 0, preorder.length - 1);
    }

    public TreeNode dfs(int[] pre, int left, int right) {
        if (left > right)
            return null;

        int val = pre[index];
        index += 1;
        TreeNode node = new TreeNode(val);
        int mid = map.get(val);
        node.left = dfs(pre, left, mid - 1);
        node.right = dfs(pre, mid + 1, right);
        return node;
    }
}